packages/daemon/src/service/trace-manager.ts

```diff
@@ -0,0 +1,250 @@
+import { Transform, TransformCallback } from 'stream';
+import { EventEmitter } from 'events';
+import { open, close, write } from 'fs-extra';
+import { provide, scope, ScopeEnum } from 'midway';
+import { StringDecoder } from 'string_decoder';
+import { generateId, PipelineStatus, PluginTypeI } from '@pipcook/pipcook-core';
+import Debug from 'debug';
+
+const debug = Debug('tracer');
+
+export type PipcookEventType = 'log' | 'job_status';
+
+/**
+ * base pipcook event, defined the fields `type` and `data`
+ */
+export class BasePipcookEvent {
+  data?: any;
+  constructor(public type: PipcookEventType) {}
+}
+
+/**
+ * pipcook event data type for job status change
+ */
+export class JobStatusChangeEvent extends BasePipcookEvent {
+  data: {
+    jobStatus: PipelineStatus;
+    step?: PluginTypeI;
+    stepAction?: 'start' | 'end';
+    queueLength?: number;
+  };
+  constructor(jobStatus: PipelineStatus, step?: PluginTypeI, stepAction?: 'start' | 'end', queueLength?: number) {
+    super('job_status');
+    this.data = { jobStatus, step, stepAction, queueLength };
+  }
+}
+
+export type LogLevel = 'info' | 'warn' | 'error';
+/**
+ * pipcook event data type for log
+ */
+export class LogEvent extends BasePipcookEvent {
+  data: {
+    level: LogLevel;
+    data: string;
+  };
+  constructor(level: LogLevel, data: string) {
+    super('log');
+    this.data = { level, data };
+  }
+}
+
+export type PipecookEvent = JobStatusChangeEvent | LogEvent;
+/**
+ * tracer for plugin installing and pipeline running.
+ * it has 2 parts: logger and event handler:
+ * logger:
+ * stdout and stderr, they are streams to pipe the logs to clients
+ * event handler:
+ * pipe the pipcook event to clients
+ */
+export class Tracer {
+  // trace id
+  id: string;
+  // stdout stream for log pipe
+  private stdout: LogPassthrough;
+  // stderr stream for log pipe
+  private stderr: LogPassthrough;
```

Comment:
Ditto, in a `Trace` object, maintain `stdout` and `stderr` looks weird a bit, extending or registration would be more reasonable.

Reply:
Why not moving this to LogEvent?

Reply:
LogEvent is a event entity, how could it handle the stream?

Reply:
Or another type or class?